### PR TITLE
Add USNG Map Collar Support

### PR DESCRIPTION
**Pull request by theduckylittle**
Adds a new "usng" layer type for rendering USNG map collars

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,7 +31,8 @@
       "react"
     ],
     "plugins": [
-      "syntax-dynamic-import"
+      "syntax-dynamic-import",
+      "transform-object-rest-spread"
```
Comment (theduckylittle) [inline package.json:35]: This is to add support for the spread syntax ( `{... stuff}`) which is used by the map collar library.

File: src/gm3/components/layers/usng.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import UsngGrid from 'usng-map-collar/UsngGrid';
+
+export const createLayer = () => {
```
Comment (theduckylittle) [inline src/gm3/components/layers/usng.js:27]: Double arrow ( `=>`) unctions are context less, aka no `this`, which when supported means they use less resources. I believe these are just transpiled to standard functions for now.

File: src/gm3/components/map.js [hunk 3]
```diff
@@ -703,25 +709,6 @@ class Map extends React.Component {
         }
     }
 
-    sortOlLayers() {
```
Comment (theduckylittle) [inline src/gm3/components/map.js:706]: This is just churn removal.

File: webpack.config.js [hunk 4]
```diff
@@ -71,7 +71,7 @@ module.exports = env => {
                     path.join(__dirname, 'node_modules/'),
                 ],
                 exclude: function(absPath) {
-                    var acceptable = ['ol', 'mapbox-to-ol-style', '@mapbox', 'jsts'];
+                    var acceptable = ['ol', 'mapbox-to-ol-style', '@mapbox', 'jsts', 'usng-map-collar'];
```
Comment (theduckylittle) [inline webpack.config.js:74]: Let's webpack know that `usng-map-collar` should be transpiled.

File: examples/desktop/mapbook.xml [hunk 5]
```diff
@@ -4,6 +4,10 @@
         The mapping services define the source of the mapping data.
     -->
 
+    <map-source name="usng-grid" type="usng">
+        <layer name="usng-grid" status="on" />
```
Comment (theduckylittle) [inline examples/desktop/mapbook.xml:8]: It may make sense to turn this off before merging as USNG will not really be globally applicable.

**Comment by brentfraser** [comment]
Map collars.  Interesting!  Was there an Issue filed related to this PR?  Is there a description of added capabilities?

**Comment by theduckylittle** [comment]
It adds the USNG map collar? I didn't have an issue for this but it was a request I received on a backchannel. Seemed like it would be a reasonable enough PR.  I can draft a ticket with a story and acceptance criteria.

**Comment by brentfraser** [comment]
I'm just curious if it means what I think it means: a margin surrounding the mapped area containing grid/graticule coordinate values.  In the interactive map?  Just the printed map?

**Comment by theduckylittle** [comment]
It's in the interactive map. You can see an example outside of GeoMoose here: 
- https://theduckylittle.github.io/usng-map-collar/

**Comment by theduckylittle** [comment]
Oh, and also in the printed map.

**Comment by brentfraser** [comment]
Ok.  Years ago I was doing similar things with Mapserver
alberta_v5 0 2_postfix_norot

**Comment by brentfraser** [comment]
So I was thinking of white margins, lat/lon, N/E labels, tick marks, etc.

**Comment by brentfraser** [comment]
All towards the "holy grail" of map rendering margin information: 

image

Not that I have a specific need; I just like good looking maps.

**Comment by klassenjs** [comment]
The lines go wonky south of the equator.

image

Just panning south a little from the above.

image

**Comment by klassenjs** [comment]
Otherwise, generally seems to work.

**Comment by theduckylittle** [comment]
That's a bug in the underlying library.

We can leave this off for now but I know there is interest in the feature.

**Comment by klassenjs** [comment]
It might just be too early in the morning for me, but what bug is in what library?

**Comment by theduckylittle** [comment]
The navigating south and causing the lines to go errant is a bug in the
usng-map-collar library.

**Comment by klassenjs** [comment]
+1 to merge from me.  

Personally, I'd leave the USNG layer enabled in the example mapbook if it doesn't cause problems.  I like the idea of showcasing what is available, and it also makes regression testing easier when the features are enabled.  Off by default would be fine.

**Comment by brentfraser** [comment]
Might be nice to add Lat/Lon and XY options too.

**Comment by klassenjs** [comment]
@brentfraser I agree, that lat/lon and xy grids would be useful to many people, but I don't think that needs to be in this PR.

USNG/MGRS ends up being a special case of grids.  It has a lot of rules around labeling (which are primarily meant for paper maps and which we technically only partially meet, but this is usable and a lot better than nothing).  It also has a lot of gotchas in implementation because it is a "grid" in multiple projections at once (all UTM zones and PS at the poles).  Most other grids are much simpler to implement (and this code could provide a template for doing so).  MGRS also has some irregular grids, which we currently ignore, around Norway where some UTM zones were extended past their natural boundaries.

Personally, I'd like to add the Maidenhead grid as a grid and for coordinate display.  It is just specially labeled 2x1 degree lon/lat boxes and subdivisions.  But that is also a separate issue from this PR.

**Comment by brentfraser** [comment]
@klassenjs  No problem, my comment was more for selecting good keywords.  I guess we could add  `<map-source name="geographic-grid" type="geographic">` and `<map-source name="xy-grid" type="xy">` etc.  in keeping with the current implementation.  Then there would be the need to specify the coordinate system.

**Comment by theduckylittle** [comment]
I'm all for merging and ticking up the lat/long grid. That's built into openlayers and a really simple add if someone wants to pick it up.

**Pull request merged**